File: frontend-admin/src/services/strategyService.ts
/**
 * Strategy Service
 * Service for managing trading strategies (CRUD operations)
 */
import { apiClient } from '@/lib/api'

// ============================================================================
// Types & Interfaces
// ============================================================================

export interface Strategy {
  id: string
  name: string
  description: string | null
  config_type: 'visual' | 'yaml' | 'pinescript'
  symbols: string[]
  timeframe: string
  is_active: boolean
  is_backtesting: boolean
  bot_id: string | null
  created_by: string | null
  config_yaml: string | null
  pinescript_source: string | null
  created_at: string
  updated_at: string
}

export interface StrategyWithStats {
  strategy: Strategy
  signals_today: number
  total_executed: number
  indicators: string[]
}

export interface StrategyWithRelations extends Strategy {
  indicators: StrategyIndicator[]
  conditions: StrategyCondition[]
}

export interface StrategyIndicator {
  id: string
  strategy_id: string
  indicator_type: string
  parameters: Record<string, any>
  order_index: number
  created_at: string
}

export interface StrategyCondition {
  id: string
  strategy_id: string
  condition_type: string
  conditions: ConditionRule[]
  logic_operator: 'AND' | 'OR'
  order_index: number
  created_at: string
}

export interface ConditionRule {
  left: string
  operator: string
  right: string
}

export interface StrategySignal {
  id: string
  strategy_id: string
  symbol: string
  signal_type: string
  entry_price: number | null
  indicator_values: Record<string, any>
  status: 'pending' | 'executed' | 'failed'
  bot_signal_id: string | null
  created_at: string
}

export interface StrategyBacktestResult {
  id: string
  strategy_id: string
  start_date: string
  end_date: string
  symbol: string
  initial_capital: number
  leverage: number
  total_trades: number
  winning_trades: number
  losing_trades: number
  win_rate: number
  profit_factor: number
  total_pnl: number
  total_pnl_percent: number
  max_drawdown: number
  sharpe_ratio: number
  created_at: string
}

export interface RunBacktestConfig {
  symbol: string
  start_date: string
  end_date: string
  initial_capital: number
  leverage: number
  margin_percent?: number
  stop_loss_percent?: number
  take_profit_percent?: number
  include_fees?: boolean
  include_slippage?: boolean
}

export interface BacktestTrade {
  entry_time: string
  exit_time: string | null
  signal_type: 'long' | 'short'
  entry_price: number
  exit_price: number | null
  quantity: number
  pnl: number | null
  pnl_percent: number | null
  exit_reason: string | null
}

export interface BacktestCandle {
  time: number  // Unix timestamp
  open: number
  high: number
  low: number
  close: number
  volume: number
}

export interface BacktestIndicatorPoint {
  time: number  // Unix timestamp
  value: number
}

export interface BacktestIndicators {
  [key: string]: BacktestIndicatorPoint[]
}

export interface BacktestEquityPoint {
  timestamp: string
  equity: number
  price: number
}

export interface BacktestRunResult {
  id: string
  strategy_id: string
  symbol: string
  start_date: string
  end_date: string
  metrics: {
    total_trades: number
    winning_trades: number
    losing_trades: number
    win_rate: number | null
    profit_factor: number | null
    total_pnl: number | null
    total_pnl_percent: number | null
    max_drawdown: number | null
    sharpe_ratio: number | null
  }
  trades: BacktestTrade[]
  equity_curve: BacktestEquityPoint[]
  candles: BacktestCandle[]
  indicators: BacktestIndicators
}

export interface CreateStrategyData {
  name: string
  description?: string
  config_type?: 'visual' | 'yaml' | 'pinescript'
  symbols?: string[]
  timeframe?: string
  bot_id?: string
  config_yaml?: string
  pinescript_source?: string
}

export interface UpdateStrategyData {
  name?: string
  description?: string
  symbols?: string[]
  timeframe?: string
  bot_id?: string
  config_yaml?: string
  pinescript_source?: string
}

export interface AddIndicatorData {
  indicator_type: string
  parameters?: Record<string, any>
  order_index?: number
}

export interface AddConditionData {
  condition_type: string
  conditions: ConditionRule[]
  logic_operator?: 'AND' | 'OR'
  order_index?: number
}

// Available indicator types
export const INDICATOR_TYPES = [
  { value: 'nadaraya_watson', label: 'Nadaraya-Watson Envelope', params: { bandwidth: 8, mult: 3.0 } },
  { value: 'rsi', label: 'RSI (Relative Strength Index)', params: { period: 14, overbought: 70, oversold: 30 } },
  { value: 'macd', label: 'MACD', params: { fast: 12, slow: 26, signal: 9 } },
  { value: 'ema', label: 'EMA (Exponential Moving Average)', params: { period: 20 } },
  { value: 'bollinger', label: 'Bollinger Bands', params: { period: 20, std_dev: 2 } },
  { value: 'atr', label: 'ATR (Average True Range)', params: { period: 14 } },
  { value: 'volume_profile', label: 'Volume Profile', params: { lookback: 24 } },
]

// Available timeframes
export const TIMEFRAMES = [
  { value: '1m', label: '1 Minuto' },
  { value: '3m', label: '3 Minutos' },
  { value: '5m', label: '5 Minutos' },
  { value: '15m', label: '15 Minutos' },
  { value: '30m', label: '30 Minutos' },
  { value: '1h', label: '1 Hora' },
  { value: '2h', label: '2 Horas' },
  { value: '4h', label: '4 Horas' },
  { value: '6h', label: '6 Horas' },
  { value: '12h', label: '12 Horas' },
  { value: '1d', label: '1 Dia' },
]

// ============================================================================
// Helper Functions
// ============================================================================

/**
 * Parse JSON string fields from API response
 */
function parseStrategyFields(data: any): any {
  if (!data) return data

  // Parse symbols if it's a string
  if (typeof data.symbols === 'string') {
    try {
      data.symbols = JSON.parse(data.symbols)
    } catch {
      data.symbols = []
    }
  }

  // Parse indicator parameters if it's a string
  if (data.parameters && typeof data.parameters === 'string') {
    try {
      data.parameters = JSON.parse(data.parameters)
    } catch {
      data.parameters = {}
    }
  }

  // Parse conditions if it's a string
  if (data.conditions && typeof data.conditions === 'string') {
    try {
      data.conditions = JSON.parse(data.conditions)
    } catch {
      data.conditions = []
    }
  }

  return data
}

// ============================================================================
// Strategy Service Class
// ============================================================================

class StrategyService {
  /**
   * Get all strategies with statistics
   */
  async getStrategies(params?: {
    limit?: number
    offset?: number
    active_only?: boolean
  }): Promise<{ strategies: StrategyWithStats[]; total: number }> {
    const response = await apiClient.instance.get('/strategies', { params })

    if (response.data?.success && response.data?.data) {
      // API returns data with strategy nested inside each item
      const strategies = response.data.data.map((item: any) => {
        // Check if strategy is nested or flat
        const strategyData = item.strategy || item
        const strategyFields = {
          id: strategyData.id,
          name: strategyData.name,
          description: strategyData.description,
          config_type: strategyData.config_type,
          symbols: strategyData.symbols,
          timeframe: strategyData.timeframe,
          is_active: strategyData.is_active,
          is_backtesting: strategyData.is_backtesting,
          bot_id: strategyData.bot_id,
          created_by: strategyData.created_by,
          config_yaml: strategyData.config_yaml,
          pinescript_source: strategyData.pinescript_source,
          created_at: strategyData.created_at,
          updated_at: strategyData.updated_at,
          documentation: strategyData.documentation,
        }

        return {
          strategy: parseStrategyFields(strategyFields),
          signals_today: item.signals_today || 0,
          total_executed: item.total_executed || 0,
          indicators: item.indicators || [],
        }
      })
      return {
        strategies,
        total: response.data.total || strategies.length
      }
    }

    return { strategies: [], total: 0 }
  }

  /**
   * Get a strategy by ID with all relations
   */
  async getStrategy(strategyId: string): Promise<StrategyWithRelations | null> {
    const response = await apiClient.instance.get(`/strategies/${strategyId}`)

    if (response.data?.success && response.data?.data) {
      const data = response.data.data

      // Parse the strategy
      const strategy = parseStrategyFields(data)

      // Parse indicators
      if (strategy.indicators) {
        strategy.indicators = strategy.indicators.map((ind: any) => parseStrategyFields(ind))
      }

      // Parse conditions
      if (strategy.conditions) {
        strategy.conditions = strategy.conditions.map((cond: any) => parseStrategyFields(cond))
      }

      return strategy
    }

    return null
  }

  /**
   * Create a new strategy
   */
  async createStrategy(data: CreateStrategyData): Promise<Strategy> {
    const response = await apiClient.instance.post('/strategies', data)

    if (response.data?.success && response.data?.data) {
      return parseStrategyFields(response.data.data)
    }

    throw new Error(response.data?.message || 'Failed to create strategy')
  }

  /**
   * Update a strategy
   */
  async updateStrategy(strategyId: string, data: UpdateStrategyData): Promise<Strategy> {
    const response = await apiClient.instance.put(`/strategies/${strategyId}`, data)

    if (response.data?.success && response.data?.data) {
      return parseStrategyFields(response.data.data)
    }

    throw new Error(response.data?.message || 'Failed to update strategy')
  }

  /**
   * Delete a strategy
   */
  async deleteStrategy(strategyId: string): Promise<void> {
    const response = await apiClient.instance.delete(`/strategies/${strategyId}`)

    if (!response.data?.success) {
      throw new Error(response.data?.message || 'Failed to delete strategy')
    }
  }

  /**
   * Activate a strategy
   */
  async activateStrategy(strategyId: string): Promise<Strategy> {
    const response = await apiClient.instance.post(`/strategies/${strategyId}/activate`)

    if (response.data?.success && response.data?.data) {
      return parseStrategyFields(response.data.data)
    }

    throw new Error(response.data?.message || 'Failed to activate strategy')
  }

  /**
   * Deactivate a strategy
   */
  async deactivateStrategy(strategyId: string): Promise<Strategy> {
    const response = await apiClient.instance.post(`/strategies/${strategyId}/deactivate`)

    if (response.data?.success && response.data?.data) {
      return parseStrategyFields(response.data.data)
    }

    throw new Error(response.data?.message || 'Failed to deactivate strategy')
  }

  /**
   * Add an indicator to a strategy
   */
  async addIndicator(strategyId: string, data: AddIndicatorData): Promise<StrategyIndicator> {
    const response = await apiClient.instance.post(`/strategies/${strategyId}/indicators`, data)

    if (response.data?.success && response.data?.data) {
      return parseStrategyFields(response.data.data)
    }

    throw new Error(response.data?.message || 'Failed to add indicator')
  }

  /**
   * Remove an indicator
   */
  async removeIndicator(strategyId: string, indicatorId: string): Promise<void> {
    const response = await apiClient.instance.delete(`/strategies/${strategyId}/indicators/${indicatorId}`)

    if (!response.data?.success) {
      throw new Error(response.data?.message || 'Failed to remove indicator')
    }
  }

  /**
   * Add a condition to a strategy
   */
  async addCondition(strategyId: string, data: AddConditionData): Promise<StrategyCondition> {
    const response = await apiClient.instance.post(`/strategies/${strategyId}/conditions`, data)

    if (response.data?.success && response.data?.data) {
      return parseStrategyFields(response.data.data)
    }

    throw new Error(response.data?.message || 'Failed to add condition')
  }

  /**
   * Remove a condition
   */
  async removeCondition(strategyId: string, conditionId: string): Promise<void> {
    const response = await apiClient.instance.delete(`/strategies/${strategyId}/conditions/${conditionId}`)

    if (!response.data?.success) {
      throw new Error(response.data?.message || 'Failed to remove condition')
    }
  }

  /**
   * Get signals for a strategy
   */
  async getSignals(strategyId: string, params?: {
    limit?: number
    offset?: number
  }): Promise<StrategySignal[]> {
    const response = await apiClient.instance.get(`/strategies/${strategyId}/signals`, { params })

    if (response.data?.success && response.data?.data) {
      return response.data.data
    }

    return []
  }

  /**
   * Run a backtest for a strategy
   */
  async runBacktest(strategyId: string, config: RunBacktestConfig): Promise<BacktestRunResult> {
    const response = await apiClient.instance.post(`/strategies/${strategyId}/backtest`, {
      symbol: config.symbol,
      start_date: config.start_date,
      end_date: config.end_date,
      initial_capital: config.initial_capital,
      leverage: config.leverage,
      margin_percent: config.margin_percent ?? 5.0,
      stop_loss_percent: config.stop_loss_percent ?? 2.0,
      take_profit_percent: config.take_profit_percent ?? 4.0,
      include_fees: config.include_fees ?? true,
      include_slippage: config.include_slippage ?? true,
    })

    if (response.data?.success && response.data?.data) {
      return response.data.data
    }

    throw new Error(response.data?.message || 'Failed to run backtest')
  }

  /**
   * Get backtest results for a strategy
   */
  async getBacktestResults(strategyId: string): Promise<StrategyBacktestResult[]> {
    const response = await apiClient.instance.get(`/strategies/${strategyId}/backtest/results`)

    if (response.data?.success && response.data?.data) {
      return response.data.data
    }

    return []
  }

  /**
   * Apply YAML configuration to a strategy
   */
  async applyYamlConfig(strategyId: string, yamlContent: string): Promise<StrategyWithRelations> {
    const response = await apiClient.instance.post(`/strategies/${strategyId}/yaml`, {
      yaml_content: yamlContent
    })

    if (response.data?.success && response.data?.data) {
      return parseStrategyFields(response.data.data)
    }

    throw new Error(response.data?.message || 'Failed to apply YAML config')
  }

  /**
   * Generate YAML template
   */
  async generateYamlTemplate(indicators?: string[], symbols?: string[]): Promise<string> {
    const response = await apiClient.instance.get('/strategies/yaml-template', {
      params: { indicators, symbols }
    })

    if (response.data?.success && response.data?.data) {
      return response.data.data.template
    }

    throw new Error('Failed to generate YAML template')
  }
}

export const strategyService = new StrategyService()

File: apps/api-python/infrastructure/services/strategy_service.py
"""
Strategy Service

Provides CRUD operations and business logic for trading strategies.
This is the main service layer for strategy management.
"""

from datetime import datetime
from decimal import Decimal
from typing import Any, Dict, List, Optional
from uuid import UUID

import structlog
import yaml

from infrastructure.database.models.strategy import (
    ConfigType,
    ConditionType,
    IndicatorType,
    LogicOperator,
    Strategy,
    StrategyBacktestResult,
    StrategyCondition,
    StrategyIndicator,
    StrategySignal,
)
from infrastructure.database.repositories.strategy import (
    StrategyBacktestResultRepository,
    StrategyConditionRepository,
    StrategyIndicatorRepository,
    StrategyRepository,
    StrategySignalRepository,
)

logger = structlog.get_logger(__name__)


class StrategyService:
    """
    Strategy Service

    Provides high-level operations for managing trading strategies:
    - Create/Update/Delete strategies
    - Configure indicators and conditions
    - Activate/Deactivate strategies
    - Parse YAML configurations
    - Link strategies to bots
    """

    def __init__(self, db_pool):
        self.db = db_pool
        self._strategy_repo = StrategyRepository(db_pool)
        self._indicator_repo = StrategyIndicatorRepository(db_pool)
        self._condition_repo = StrategyConditionRepository(db_pool)
        self._signal_repo = StrategySignalRepository(db_pool)
        self._backtest_repo = StrategyBacktestResultRepository(db_pool)

    # ==================== Strategy CRUD ====================

    async def create_strategy(
        self,
        name: str,
        description: Optional[str] = None,
        config_type: ConfigType = ConfigType.VISUAL,
        symbols: List[str] = None,
        timeframe: str = "5m",
        bot_id: Optional[str] = None,
        created_by: Optional[str] = None,
        config_yaml: Optional[str] = None,
        pinescript_source: Optional[str] = None,
    ) -> Strategy:
        """
        Create a new trading strategy

        Args:
            name: Strategy name
            description: Optional description
            config_type: Configuration type (visual, yaml, pinescript)
            symbols: List of trading symbols (e.g., ["BTCUSDT", "ETHUSDT"])
            timeframe: Trading timeframe (1m, 5m, 15m, 1h, etc.)
            bot_id: Optional linked bot ID
            created_by: User ID who created the strategy
            config_yaml: YAML configuration (if config_type is YAML)
            pinescript_source: PineScript source (if config_type is PINESCRIPT)

        Returns:
            Created Strategy object
        """
        strategy = Strategy(
            name=name,
            description=description,
            config_type=config_type,
            symbols=symbols or [],
            timeframe=timeframe,
            bot_id=bot_id,
            created_by=created_by,
            config_yaml=config_yaml,
            pinescript_source=pinescript_source,
            is_active=False,
            is_backtesting=False,
        )

        created = await self._strategy_repo.create_strategy(strategy)

        # If YAML config provided, parse and create indicators/conditions
        if config_type == ConfigType.YAML and config_yaml:
            await self._apply_yaml_config(str(created.id), config_yaml)

        logger.info(
            f"Strategy created: {name}",
            strategy_id=str(created.id),
            symbols=symbols
        )

        return created

    async def get_strategy(self, strategy_id: str) -> Optional[Strategy]:
        """Get a strategy by ID with all relations loaded"""
        return await self._strategy_repo.get_with_relations(strategy_id)

    async def list_strategies(
        self,
        user_id: Optional[str] = None,
        active_only: bool = False,
        limit: int = 50,
        offset: int = 0,
    ) -> List[Dict[str, Any]]:
        """
        List strategies with statistics

        Args:
            user_id: Filter by creator user ID
            active_only: Only return active strategies
            limit: Maximum number of results
            offset: Pagination offset

        Returns:
            List of strategies with signal statistics
        """
        return await self._strategy_repo.get_strategies_with_stats(
            user_id=user_id,
            active_only=active_only,
            limit=limit,
            offset=offset
        )

    async def update_strategy(
        self,
        strategy_id: str,
        **updates
    ) -> Optional[Strategy]:
        """
        Update a strategy

        Args:
            strategy_id: Strategy ID
            **updates: Fields to update

        Returns:
            Updated Strategy object
        """
        # Prevent updating certain fields
        updates.pop("id", None)
        updates.pop("created_at", None)

        updated = await self._strategy_repo.update(strategy_id, **updates)

        if updated:
            logger.info(f"Strategy updated: {strategy_id}", updates=list(updates.keys()))

        return updated

    async def delete_strategy(self, strategy_id: str) -> bool:
        """
        Delete a strategy (soft delete - marks as inactive first)

        Args:
            strategy_id: Strategy ID

        Returns:
            True if deleted
        """
        # Deactivate first
        await self._strategy_repo.update(strategy_id, is_active=False)

        # Then delete
        deleted = await self._strategy_repo.delete(strategy_id)

        if deleted:
            logger.info(f"Strategy deleted: {strategy_id}")

        return deleted

    # ==================== Activation ====================

    async def activate_strategy(self, strategy_id: str) -> Strategy:
        """
        Activate a strategy for live trading

        Args:
            strategy_id: Strategy ID

        Returns:
            Updated Strategy
        """
        strategy = await self._strategy_repo.get(strategy_id)
        if not strategy:
            raise ValueError(f"Strategy {strategy_id} not found")

        # Validate strategy has required configuration
        full_strategy = await self._strategy_repo.get_with_relations(strategy_id)

        if not full_strategy.get_symbols_list():
            raise ValueError("Strategy must have at least one symbol configured")

        if not full_strategy.indicators:
            raise ValueError("Strategy must have at least one indicator configured")

        # Activate
        updated = await self._strategy_repo.update(strategy_id, is_active=True)

        logger.info(f"Strategy activated: {strategy_id}")

        return updated

    async def deactivate_strategy(self, strategy_id: str) -> Strategy:
        """
        Deactivate a strategy

        Args:
            strategy_id: Strategy ID

        Returns:
            Updated Strategy
        """
        updated = await self._strategy_repo.update(strategy_id, is_active=False)

        logger.info(f"Strategy deactivated: {strategy_id}")

        return updated

    # ==================== Indicators ====================

    async def add_indicator(
        self,
        strategy_id: str,
        indicator_type: IndicatorType,
        parameters: Dict[str, Any] = None,
        order_index: int = 0,
    ) -> StrategyIndicator:
        """
        Add an indicator to a strategy

        Args:
            strategy_id: Strategy ID
            indicator_type: Type of indicator
            parameters: Indicator parameters
            order_index: Order for evaluation

        Returns:
            Created StrategyIndicator
        """
        indicator = StrategyIndicator(
            strategy_id=strategy_id,
            indicator_type=indicator_type,
            parameters=parameters or {},
            order_index=order_index,
        )

        created = await self._indicator_repo.create_indicator(indicator)

        logger.info(
            f"Indicator added",
            strategy_id=strategy_id,
            indicator_type=indicator_type.value
        )

        return created

    async def update_indicator(
        self,
        indicator_id: str,
        **updates
    ) -> Optional[StrategyIndicator]:
        """Update an indicator's parameters"""
        return await self._indicator_repo.update(indicator_id, **updates)

    async def remove_indicator(self, indicator_id: str) -> bool:
        """Remove an indicator from a strategy"""
        return await self._indicator_repo.delete(indicator_id)

    async def get_indicators(self, strategy_id: str) -> List[StrategyIndicator]:
        """Get all indicators for a strategy"""
        return await self._indicator_repo.get_by_strategy(strategy_id)

    # ==================== Conditions ====================

    async def add_condition(
        self,
        strategy_id: str,
        condition_type: ConditionType,
        conditions: List[Dict[str, Any]],
        logic_operator: LogicOperator = LogicOperator.AND,
        order_index: int = 0,
    ) -> StrategyCondition:
        """
        Add a condition to a strategy

        Args:
            strategy_id: Strategy ID
            condition_type: Type (entry_long, entry_short, exit_long, exit_short)
            conditions: List of condition rules
            logic_operator: AND or OR
            order_index: Order for evaluation

        Returns:
            Created StrategyCondition

        Example conditions:
            [
                {"left": "close", "operator": "<", "right": "ndy.lower"},
                {"left": "close", "operator": "<", "right": "tpo.val"}
            ]
        """
        condition = StrategyCondition(
            strategy_id=strategy_id,
            condition_type=condition_type,
            conditions=conditions,
            logic_operator=logic_operator,
            order_index=order_index,
        )

        created = await self._condition_repo.create_condition(condition)

        logger.info(
            f"Condition added",
            strategy_id=strategy_id,
            condition_type=condition_type.value
        )

        return created

    async def update_condition(
        self,
        condition_id: str,
        **updates
    ) -> Optional[StrategyCondition]:
        """Update a condition"""
        return await self._condition_repo.update(condition_id, **updates)

    async def remove_condition(self, condition_id: str) -> bool:
        """Remove a condition from a strategy"""
        return await self._condition_repo.delete(condition_id)

    async def get_conditions(self, strategy_id: str) -> List[StrategyCondition]:
        """Get all conditions for a strategy"""
        return await self._condition_repo.get_by_strategy(strategy_id)

    # ==================== Signals ====================

    async def get_signals(
        self,
        strategy_id: str,
        limit: int = 100,
        offset: int = 0,
    ) -> List[StrategySignal]:
        """Get signals generated by a strategy"""
        return await self._signal_repo.get_by_strategy(strategy_id, limit=limit, offset=offset)

    async def get_recent_signals(
        self,
        limit: int = 50,
        symbol: Optional[str] = None,
    ) -> List[StrategySignal]:
        """Get recent signals across all strategies"""
        return await self._signal_repo.get_recent(limit=limit, symbol=symbol)

    # ==================== Bot Linking ====================

    async def link_bot(self, strategy_id: str, bot_id: str) -> Strategy:
        """
        Link a strategy to a bot for signal execution

        Args:
            strategy_id: Strategy ID
            bot_id: Bot ID to link

        Returns:
            Updated Strategy
        """
        # Verify bot exists
        bot = await self.db.fetchrow(
            "SELECT id, name FROM bots WHERE id = $1",
            UUID(bot_id)
        )
        if not bot:
            raise ValueError(f"Bot {bot_id} not found")

        updated = await self._strategy_repo.update(strategy_id, bot_id=bot_id)

        logger.info(
            f"Strategy linked to bot",
            strategy_id=strategy_id,
            bot_id=bot_id,
            bot_name=bot["name"]
        )

        return updated

    async def unlink_bot(self, strategy_id: str) -> Strategy:
        """
        Unlink a strategy from its bot

        Args:
            strategy_id: Strategy ID

        Returns:
            Updated Strategy
        """
        updated = await self._strategy_repo.update(strategy_id, bot_id=None)

        logger.info(f"Strategy unlinked from bot", strategy_id=strategy_id)

        return updated

    # ==================== YAML Configuration ====================

    async def apply_yaml_config(self, strategy_id: str, yaml_content: str) -> Strategy:
        """
        Apply YAML configuration to a strategy

        Args:
            strategy_id: Strategy ID
            yaml_content: YAML configuration string

        Returns:
            Updated Strategy
        """
        # Validate and apply
        await self._apply_yaml_config(strategy_id, yaml_content)

        # Update strategy with YAML
        updated = await self._strategy_repo.update(
            strategy_id,
            config_type=ConfigType.YAML,
            config_yaml=yaml_content
        )

        logger.info(f"YAML config applied", strategy_id=strategy_id)

        return updated

    async def _apply_yaml_config(self, strategy_id: str, yaml_content: str) -> None:
        """Parse YAML and create indicators/conditions"""
        try:
            config = yaml.safe_load(yaml_content)
        except yaml.YAMLError as e:
            raise ValueError(f"Invalid YAML: {e}")

        # Clear existing indicators and conditions
        existing_indicators = await self._indicator_repo.get_by_strategy(strategy_id)
        for ind in existing_indicators:
            await self._indicator_repo.delete(str(ind.id))

        existing_conditions = await self._condition_repo.get_by_strategy(strategy_id)
        for cond in existing_conditions:
            await self._condition_repo.delete(str(cond.id))

        # Update symbols and timeframe if provided
        if "symbols" in config:
            await self._strategy_repo.update(strategy_id, symbols=config["symbols"])
        if "timeframe" in config:
            await self._strategy_repo.update(strategy_id, timeframe=config["timeframe"])

        # Create indicators
        indicators_config = config.get("indicators", [])
        for idx, ind_config in enumerate(indicators_config):
            ind_type = ind_config.get("type", "")
            try:
                indicator_type = IndicatorType(ind_type)
            except ValueError:
                logger.warning(f"Unknown indicator type: {ind_type}")
                continue

            await self.add_indicator(
                strategy_id=strategy_id,
                indicator_type=indicator_type,
                parameters=ind_config.get("parameters", {}),
                order_index=idx,
            )

        # Create conditions
        conditions_config = config.get("conditions", {})

        for cond_type_str, cond_list in conditions_config.items():
            try:
                condition_type = ConditionType(cond_type_str)
            except ValueError:
                logger.warning(f"Unknown condition type: {cond_type_str}")
                continue

            logic = cond_list.get("logic", "AND")
            logic_operator = LogicOperator.AND if logic.upper() == "AND" else LogicOperator.OR

            rules = cond_list.get("rules", [])

            await self.add_condition(
                strategy_id=strategy_id,
                condition_type=condition_type,
                conditions=rules,
                logic_operator=logic_operator,
            )

    def generate_yaml_template(
        self,
        indicator_types: List[str] = None,
        symbols: List[str] = None,
    ) -> str:
        """
        Generate a YAML template for strategy configuration

        Args:
            indicator_types: List of indicator types to include
            symbols: List of symbols

        Returns:
            YAML template string
        """
        template = {
            "name": "My Strategy",
            "description": "Strategy description",
            "symbols": symbols or ["BTCUSDT"],
            "timeframe": "5m",
            "indicators": [],
            "conditions": {
                "entry_long": {
                    "logic": "AND",
                    "rules": []
                },
                "entry_short": {
                    "logic": "AND",
                    "rules": []
                }
            }
        }

        # Add indicator templates
        indicator_templates = {
            "nadaraya_watson": {
                "type": "nadaraya_watson",
                "parameters": {
                    "bandwidth": 8,
                    "mult": 3.0,
                    "src": "close",
                    "use_atr": True,
                    "atr_period": 14
                }
            },
            "tpo": {
                "type": "tpo",
                "parameters": {
                    "period": 24,
                    "value_area_percent": 70.0,
                    "row_size": 24
                }
            },
            "rsi": {
                "type": "rsi",
                "parameters": {
                    "period": 14,
                    "overbought": 70,
                    "oversold": 30
                }
            }
        }

        if indicator_types:
            for ind_type in indicator_types:
                if ind_type in indicator_templates:
                    template["indicators"].append(indicator_templates[ind_type])
        else:
            # Default: NDY + TPO
            template["indicators"] = [
                indicator_templates["nadaraya_watson"],
                indicator_templates["tpo"]
            ]

        # Add condition templates
        template["conditions"]["entry_long"]["rules"] = [
            {"left": "close", "operator": "<", "right": "nadaraya_watson.lower"},
            {"left": "close", "operator": "<", "right": "tpo.val"}
        ]

        template["conditions"]["entry_short"]["rules"] = [
            {"left": "close", "operator": ">", "right": "nadaraya_watson.upper"},
            {"left": "close", "operator": ">", "right": "tpo.vah"}
        ]

        return yaml.dump(template, default_flow_style=False, sort_keys=False)

    # ==================== Backtest Results ====================

    async def get_backtest_results(
        self,
        strategy_id: str,
        limit: int = 10,
    ) -> List[StrategyBacktestResult]:
        """Get backtest results for a strategy"""
        return await self._backtest_repo.get_by_strategy(strategy_id, limit=limit)

    async def get_latest_backtest(
        self,
        strategy_id: str,
    ) -> Optional[StrategyBacktestResult]:
        """Get the latest backtest result for a strategy"""
        return await self._backtest_repo.get_latest(strategy_id)
